### ISSUE TITLE
Server vs Client checksums (and deprecate my own project)?

### ISSUE THREAD

**Issue by ctrlplusb** [thejameskyle/react-loadable#6]
Hey James;

Great project!  I really like some of your choices for the API to keep things as simple and powerful as possible.  I actually have a very very very similar effort: [`react-async-component`](https://github.com/thejameskyle/react-loadable).

It's actually so similar that I would be happy to deprecate it and redirect everyone to your project as it seems to have a lot of the community behind it already. 😀

One thing I'd like to query with you before I seriously consider doing so: how are you managing any mismatch between the server side render checksum vs the initial client side render checksum?  In my personal experience, as the client loads the component asynchronously the initial render of the client never matches what is returned by the server. This causes checksum failures and React therefore tries to re-render the client over the DOM (causing a flash of content). 

To resolve this I had to introduce some witch-craft in my own implementation so that I could track and then prefetch any of the required webpack chunks on the client (that were loaded for a server request) before then doing the client render. It's all fairly complicated compared to the rest of it, and perhaps you have already consider this and decided it's not worth the complexity.

Otherwise in terms of feature set there is very little to distinguish our two libraries.  You have the delay prop, which is rad and effective.  I have an "ssrMode" flag which allows you to control render behaviour for the component on the server ("defer", "boundary", "render" - my docs describe them in more detail).

I'd be interested to hear your thoughts, and like I said I am more than happy to shut down my shop and instead contribute to your own library.

**Comment by thejameskyle** [comment]
Hey Sean, thanks for opening this up.

So the solution has several steps in the future.

1. I need to work with @TheLarkInn on some webpack runtime features to enable this sort of server to work https://gist.github.com/thejameskyle/abecfe8ec2a7ce1e312a904527a31908
2. I've been experimenting with tooling that abstracts away this problem (I haven't gotten very far: https://github.com/thejameskyle/packager)
3. I'm going to try figuring out some hooks that webpack-dev-server can add to enable the work I do in 2
4. Once React Fiber lands and is adopted we can solve the problem even better because we can defer work until bundles are available. I'm going to keep picking at @sebmarkbage to figure out what we can and should do.

That's all pretty vague, but I think it'll work out in the end. React Loadable isn't really a necessary piece to any of this, but it helped me think about this path forward for the community.

**Comment by thejameskyle** [comment]
Also, we're in a weird situation right now with server-side rendering. I would be hesitant to suggest that anyone use server-side rendering at this point. I just want to enable it well enough that people adopt patterns that are going to work better once we have nicer tooling.

**Comment by ctrlplusb** [comment]
Awesome, it's really exciting to hear that the big players of the React world are focusing on this now.  I'm super excited for what you all come up with.  If you ever need someone to test out anything you come up with I would be more than happy to oblige. 😀

**Comment by gogoyqj** [comment]
great project @thejameskyle 

I tried to replace `import a Component` with a promise, and make changes to navigator[router] to ensure all children components resolved before root component's setState/rerender triggering. seems not a really good idea.

react-loadable made a new way.

i run the [exmaple](https://github.com/thejameskyle/react-loadable-example), seems that lifecycle changed.

**Comment by smashercosmo** [comment]
I think that one of the easiest way is to keep components in the some object provided through the react context. In this case we can make to renders on the client. First one to just trigger componentWillMount hooks, and the second one to actually render app. After first render pass we can loop through component promises and then after all components are loaded we can render our application. I have a proof of concept here https://github.com/smashercosmo/react-stack-playground.

**Comment by ctrlplusb** [comment]
@smashercosmo yeah, thats kinda what mine does.

**Comment by simenbrekken** [comment]
@smashercosmo I've implemented something similar that handles recursive dependencies as well: https://github.com/unfold/fetch-data/blob/master/src/resolver.js

**Comment by faceyspacey** [comment]
@thejameskyle would it be accurate to sum the problem up as being there needs to be a `hasModule()` function client side (provided by webpack) and a way to toggle a synchronous `require` if so, *but without triggering* webpack to also include that module in the bundle (the idea being it's a code split bundle already embedded on the page by virtue of correct server-side rendering). 

All the rage has been how code-splitting is solved now, especially thanks to `import().next`, but it's my opinion until we take it all the way as just described--and I think what @ctrlplusb was originally getting at--it's just a toy from which you can get perhaps 15-25% the value you will get once the entire intent is complete. I use code splitting for pages like Privacy Policy, modals, and other occasionally visited static pages. The idea being if you visit them correctly, there's additional latency because of the additional request to get the split module, while giving the benefit of file size reducer of your main bundle. 

So basically it's currently only a tool to section off parts of your app you don't want taking up bytes from more popular pages within your app, whereas the dream is a new paradigm where you're code-splitting willy-nilly everywhere without much thought necessary. For now requiring synchronously won't work because of the checksum mismatch and therefore you'll get a re-render and a longer time till interactive. And if you really do want to pull off a syncronous require client side of a code-split package, it requires lots of "witch craft" that certainly doesn't fit into a neat reusable pattern as does *React Loadable* as a component.

**Comment by thejameskyle** [comment]
@faceyspacey I don't follow, and please don't rant at me
